pkg/liveness/liveness.go

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2018 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package liveness
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+var felixLivenessEp string
+
+func init() {
+	felixPort := os.Getenv("FELIX_HEALTHPORT")
+	if felixPort == "" {
+		felixLivenessEp = "http://localhost:9099/liveness"
+		return
+	}
+
+	if _, err := strconv.Atoi(felixPort); err != nil {
+		log.Panicf("Failed to parse value for port %s", felixPort)
+	}
+
+	felixLivenessEp = "http://localhost:" + felixPort + "/liveness"
+}
+
+// CheckFelixLive Checks whether or not felix is live.
+func CheckFelixLive() (err error) {
+	c := &http.Client{Timeout: 5 * time.Second}
+	resp, err := c.Get(felixLivenessEp)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if e := resp.Body.Close(); e != nil {
```

Comment:
I wonder if this whole defer block can be reduced to:

```go
defer func() {
     err = resp.Body.Close()
}
```

Reply:
I don't think it can.  In the case that `Close()` returns `nil` but the status code is bad, I think we'd hit this line:
```
return fmt.Errorf("liveness probe reporting %d", resp.StatusCode)
```
and then overwrite `err` with `nil` when the defer runs, suppressing the error.

Reply:
Right... I over-optimized the defer.